hie-plugin-api/Haskell/Ide/Engine/PluginDescriptor.hs

```diff
@@ -214,7 +233,8 @@ data ParamVal (t :: ParamType) where
 class ValidResponse a where
   jsWrite :: a -> Object -- ^ Serialize to JSON Object
   jsRead  :: Object -> Parser a -- ^ Read from JSON Object
-
+  responseTypeName :: a -> T.Text -- ^ Type name in command description.
```

Comment:
I explicitly put a specific type name for two reasons:
- be able to have several types act as one in the API (String and Text, for example)
- have one layer of indirection between the Haskell types and the types we expose, to avoid leaking our Haskell into our JSON (see what I fixed earlier on parameters RP/OP)

But of course the flip side is that it's easy to define two different structures that return the same responseTypeName with different JSON, so I'll agree we could use TypeRep and avoid types like String.
@alanz  I don't understand your comment about the type being clearly serialized. For me, an IDE needs to know the return type of a command BEFORE it even offers it to the user, because potentially it needs to know what to do with the response. For example, a refactor operation changes the files on disk and return the paths of the change files. the IDE needs to understand that, so that it can reload the files. A generic method that will offer the command to the user without doing anything with the answer will cause issues (the IDE file representation being out of sync with the disk). But at the same time, if I know how to handle a specific return type, I can handle any command that return that type. So you can add loads of Hare refactor commands, if they all return the same type I know the IDE can support them all, without a priori knowledge of which commands to support.